Use external library for date handling code

chrismytton opened the pull request:
This uses the [date_range](https://github.com/everypolitician/date_range) for handling date overlapping. I've taken this opportunity to refactor things into a `Membership` class, which encapsulates the date overlapping logic.

Fixes https://github.com/everypolitician/combine_popolo_memberships/issues/8

chrismytton committed:
Add date-range dependency

This will let us simplify the logic for checking of two dates overlap
and if so getting the correct start and end dates for the overlap.

chrismytton committed:
Refactor overlap logic into Membership class

This introduces a new Membership class which uses `date-range` to
calculate the correct overlap with another membership.

tmtmtmtm commented on lib/combine_popolo_memberships.rb:
```diff
@@ -1,30 +1,50 @@
+    end
+
+    def to_h
+      raw
```
A public method that does nothing but expose a private method implies that something odd is going on somewhere…

chrismytton replied:
Yeah, not sure what's best to do about this. For now I've just exposed `membership_hash` (the new name for `raw`) in the public interface and dropped this `#to_h` method. 7d39d37

tmtmtmtm commented on lib/combine_popolo_memberships.rb:
```diff
@@ -1,30 +1,50 @@
+    def overlap(other)
+      o = date_range.overlap(other.date_range)
+      return unless o
+      {}.merge(to_h).merge(other.to_h).tap do |h|
```
This is a quite confusing construction. Why not just `to_h.merge`?

chrismytton replied:
Ha, oops. Fixed in b860618.

tmtmtmtm commented on test/membership_test.rb:
```diff
@@ -0,0 +1,36 @@
+
+  def end_date
+    '2015-12-31'
+  end
```
Each of these are only used in a single place, so it's a bit odd to create methods for them…

chrismytton replied:
Yep, I think they're a hangover from where I coped them from #2. Have removed them in 609e56f.

tmtmtmtm commented on test/membership_test.rb:
```diff
@@ -0,0 +1,36 @@
+    membership = CombinePopoloMemberships::Membership.new(start_date: '', end_date: '')
+    assert_equal({ start_date: '', end_date: '' }, membership.to_h)
+    membership2 = CombinePopoloMemberships::Membership.new(start_date: nil, end_date: nil)
+    assert_equal({ start_date: nil, end_date: nil }, membership2.to_h)
```
I think it would be clearer if the `to_h` tests were included with each of earlier groups, rather than in a separate method of its own.

chrismytton replied:
Ah yep, done in 167c66d.

tmtmtmtm commented on lib/combine_popolo_memberships.rb:
```diff
@@ -1,30 +1,50 @@
+      @raw = raw
+    end
+
+    def start_date
```
I fear that this class has a bit too much in its public interface. It looks like `start_date` and `end_date` are only there so you can access them in the test? If so, I'd suggest making them private, and changing the test to get at the through the range instead.

chrismytton replied:
Good idea. I've switched to getting the dates through `date_range` in the tests in bfc873d and then made `start_date` and `end_date` private in 4dd049d.

tmtmtmtm commented on lib/combine_popolo_memberships.rb:
```diff
@@ -1,30 +1,50 @@
-      end_date:   e == '9999-99-99' ? nil : e
-    }
+  class Membership
+    def initialize(raw)
```
`raw` is a slightly awkward name here, especially as there's no documentation. It's not immediately obvious what this actually *is*.

chrismytton replied:
Agreed. I've changed it to `membership_hash` in b2a8a50.

tmtmtmtm commented on lib/combine_popolo_memberships.rb:
```diff
@@ -1,30 +1,50 @@
+      h = Membership.new(from).overlap(Membership.new(to))
+      next unless h
+      h.delete(:id)
       h.merge(from_name => from[:id], into_name => to[:id])
```
it seems a little odd to still be accessing data directly from `from` and `to` here, when you've extracted a `Membership` to encapsulate them.

chrismytton replied:
The reason I've left this code alone for now is we're going to change the interface anyway when we fix #7, so I don't want to rush into anything now and add a `Membership#id` method which we might not need for the new interface.

So I think it's best to leave this as-is for now, then we can refactor this code in a future PR.

tmtmtmtm replied:
I don't really follow your thinking here, I'm afraid. The `Membership` class is essentially a private implementation — it doesn't _have_ a public interface. The only reason it seems to be exposed at all is so that you can unit test it (which I'm not sure has a lot of value in its own right, and definitely doesn't if it's already making things this rigid.)

It's not clear to me that that the `Membership` class will ever be directly part of the public interface — it's certainly not obvious so far that it will, though perhaps you've already thought through, or even have WIP code for, the next step. (If that _is_ the plan, then it worries me slightly that `Membership` may not be the right abstraction here, as one of the common usage patterns here is for one of the data lists to be a list of terms, which aren't really memberships)

tmtmtmtm reviewed (changes requested):
This was quite hard to follow, which on the one hand is understandable as the underlying code already is, but I fear that some of this is in the uncanny valley of being either a bit too big of a step (rather than simply switching in the library, and then refactoring separately), or too small of one. 

There are definitely places here where I'd be a lot more comfortable if we had a more comprehensive test-suite. I wonder if it's worth bulking that out a bit more separately, with more real-world examples from some of the places we've been using it (e.g. the double-step combination we're doing with the Dutch Wikidata)?

chrismytton commented:
> There are definitely places here where I'd be a lot more comfortable if we had a more comprehensive test-suite. I wonder if it's worth bulking that out a bit more separately, with more real-world examples from some of the places we've been using it (e.g. the double-step combination we're doing with the Dutch Wikidata)?

Yeah, I agree. I'll add some more tests for the existing code in a separate PR, then come back to this and do the smaller step of swapping in date-range as a separate change.

chrismytton committed:
Give Membership constructor argument a better name

This changes the constructor argument variable from `raw` to
`membership_hash`, which should make it much clearer what the argument
is expected to contain.

chrismytton committed:
Simplify merge code in Membership#overlap

chrismytton committed:
Make Membership#membership_hash public

Rather than the indirection of having a public method expose a private
method just make the method public in the first place.

chrismytton committed:
Remove unnecessary methods in tests

chrismytton committed:
Use date_range to get to dates in tests

This means we can have a smaller public interface on the membership
because we won't have to expose the start date and end date methods.

chrismytton committed:
Make Membership start and end dates private

These don't need to be in the public interface as they can be accessed
through the date_range.

chrismytton committed:
Combine membership_hash tests with date_range tests

Rather than having to setup the same Membership objects twice, just run
all the tests we need to at once.

chrismytton commented:
> This was quite hard to follow, which on the one hand is understandable as the underlying code already is, but I fear that some of this is in the uncanny valley of being either a bit too big of a step (rather than simply switching in the library, and then refactoring separately), or too small of one.

I don't think this was too big a step. The problem with just switching in the library and then refactoring separately is that the library isn't a drop-in replacement. We first need to get rid of the awkwardness of passing the two memberships back in the `:_data` key from `.overlap` in order to merge them together.

> There are definitely places here where I'd be a lot more comfortable if we had a more comprehensive test-suite. I wonder if it's worth bulking that out a bit more separately, with more real-world examples from some of the places we've been using it (e.g. the double-step combination we're doing with the Dutch Wikidata)?

I started writing tests for some real world data, but quickly realised that I was testing the `CombinePopoloMemberships.combine` interface, which is going to change soon when we fix #7. So I don't want to write loads of tests, only for us to decide that the output hash should be merged differently. The other option was to test `CombinePopoloMemberships.overlap`, but that functionality is going to be replaced with a combination of [`date-range`](https://github.com/everypolitician/date_range), which has its own test suite, and the `Membership` class in this pull request.

So I think the best thing to do is to get this merged, then we can write some more comprehensive tests for the `Membership` class and update the interface as enhancements on top of this change.

tmtmtmtm commented:
> I started writing tests for some real world data, but quickly realised that I was testing the `CombinePopoloMemberships.combine` interface, which is going to change soon when we fix #7. So I don't want to write loads of tests, only for us to decide that the output hash should be merged differently. The other option was to test `CombinePopoloMemberships.overlap`, but that functionality is going to be replaced with a combination of [`date-range`](https://github.com/everypolitician/date_range), which has its own test suite, and the `Membership` class in this pull request.

I think you've talked yourself into a false dichotomy here. The tests should be about making sure that, given a series of data hashes, the code slices and dices and puts them back again correctly — not just testing specific methods. If we're going to be changing the public interface anyway, then that makes the tests *more* valuable, not less. If you want to minimise how many changes will be needed to the tests when that happens, you could make the tests be largely data-driven, so that the code that interacts directly with the class is encapsulated into one place. But even without that, this shouldn't really be a problem. Any changes to the interface are presumably going to be done in a backwards compatible manner (e.g. adding a new interface (and testing that); then making the old interface delegate transparently to that (with the existing regression tests as the safety net that that doesn't change anything), then (later) dropping the old interface.

> So I think the best thing to do is to get this merged, then we can write some more comprehensive tests for the `Membership` class and update the interface as enhancements on top of this change.

Even if the code worked I'd be quite hesitant to do that, but I tried swapping this in to [the New Zealand scraper](https://github.com/everypolitician-scrapers/new-zealand-parliament), and it breaks quite badly.

For example, when dealing with Jami-Lee Ross, the old version ended up with these data hashes:

```ruby
{:start_date=>"2011-03-05", :end_date=>"2011-10-20", :name=>"49th Parliament", :wikidata=>"Q11812120", :area=>"Botany", :party=>"National Party", :term=>"49", :id=>"51MP3101"}
{:start_date=>"2011-12-20", :end_date=>"2014-09-19", :name=>"50th Parliament", :wikidata=>"Q4120749", :area=>"Botany", :party=>"National Party", :term=>"50", :id=>"51MP3101"}
{:start_date=>"2014-10-20", :end_date=>nil, :name=>"51st Parliament", :wikidata=>"Q4640115", :area=>"Botany", :party=>"National Party", :term=>"51", :id=>"51MP3101"}
```

whereas the new code gives:

```ruby
{:name=>"49th Parliament", :start_date=>"2011-03-05", :end_date=>"2011-10-20", :wikidata=>"Q11812120", :area=>"Botany", :party=>"National Party", :term=>"49", :id=>"51MP3101"}
{:name=>"50th Parliament", :start_date=>"2011-12-20", :end_date=>"2014-09-19", :wikidata=>"Q4120749", :area=>"Botany", :party=>"National Party", :term=>"50", :id=>"51MP3101"}
{:name=>"51st Parliament", :start_date=>"2014-10-20", :end_date=>"9999-99-99", :wikidata=>"Q4640115", :area=>"Botany", :party=>"National Party", :term=>"51", :id=>"51MP3101"}
```
The extra `name` field isn't an issue in this specific case, as we [explicitly filter the fields we want](https://github.com/everypolitician-scrapers/new-zealand-parliament/blob/bbeb251/scraper.rb#L36-L37), but it presumably would be in other cases, and leaking out the '9999-99-99' end date is a problem.

tmtmtmtm commented:
> I don't think this was too big a step. The problem with just switching in the library and then refactoring separately is that the library isn't a drop-in replacement. 

Even if it's not a drop-in replacement, I'm still concerned that this PR is an awkward half-way house. Conceptually it's meant to have no change in user-facing behaviour — i.e. it's a refactoring. Yes, we would like to make the interface more intuitive, but *this* PR isn't about that — this is meant to just be simplifying the implementation. But right now I'm not convinced that it's doing that — it's introducing a new half-public-half-private class, that doesn't really hide the data it's encapsulating and may have a misleading name that will cause even more confusion later. Even in its current form I'm far from sure that this is an improvement, and, of course, if it doesn't actually work right yet, it's not clear what the final form will end up as.

I'm worried that fixing all this will burn up even more time and effort, when this isn't really on the critical path to anything. Unless I've forgotten a key part of why we're doing this work, the library already works well enough as is to do what we need. Tidying things up would definitely be nice, but is it really something we can spend weeks or even days on right now? Or, if we're doing this to enable something in particular, then we might need to rethink what's actually needed to get to that point.
